In [2]:
!pip install seaborn 
!pip install sklearn
!pip install statsmodels

  Running setup.py bdist_wheel for sklearn ... - done
  Stored in directory: /root/.cache/pip/wheels/76/03/bb/589d421d27431bcd2c6da284d5f2286c8e3b2ea3cf1594c074
Successfully built sklearn


In [3]:
import os 
import pandas as pd
import numpy as np 
import seaborn as sns
%matplotlib inline

from sklearn.model_selection import train_test_split
from patsy import dmatrices
import statsmodels.api as sm
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix 

/usr/local/lib/python3.6/dist-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


## Reading in Dataset

In [0]:
# reading in Dataset
paths = 'http://bigdata.cs.byu.edu/Lectures/EEGEyeState.csv'
df = pd.read_table(paths,sep=',',encoding='latin1')

# split dependent and independent variables
X = list(df)[:-1]
y = ['eyeDetection']

# split into dataframe
y, X = dmatrices('eyeDetection~AF3+F7+F3+FC5+T7+P7+O1+O2+P8+T8+FC6+F4+F8+AF4',df,return_type='dataframe') 

# create test and train.
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=.1,random_state=40)

# Random Forest

In [4]:
rf = RandomForestClassifier()
param = {'n_estimators': [100, 150, 250],
        'max_depth': [10, 30, 50]}

gs = GridSearchCV(rf, param, cv=5, n_jobs=-1)
gs_fit = gs.fit(X_train, y_train)
pd.DataFrame(gs_fit.cv_results_).sort_values('mean_test_score', ascending=False)

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:458: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:458: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:458: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:458: DataConversionWarning: A column-vector y was passed when a 

mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
5       8.140881         0.183678         0.929016          1.000000   
8       7.898094         0.176884         0.928275          1.000000   
7       4.875845         0.109060         0.927014          1.000000   
4       4.901674         0.111024         0.926940          1.000000   
3       3.258447         0.075768         0.925605          1.000000   
6       3.261936         0.073309         0.924714          1.000000   
1       3.496710         0.079272         0.860110          0.914274   
0       2.075374         0.050490         0.859516          0.911382   
2       5.752641         0.133264         0.858923          0.913329   

  param_max_depth param_n_estimators                                  params  \
5              30                250  {'max_depth': 30, 'n_estimators': 250}   
8              50                250  {'max_depth': 50, 'n_estimators': 250}   
7              50                150  {'max_depth': 50, 'n_estimators': 150}   
4              30                150  {'max_depth': 30, 'n_estimators': 150}   
3              30                100  {'max_depth': 30, 'n_estimators': 100}   
6              50                100  {'max_depth': 50, 'n_estimators': 100}   
1              10                150  {'max_depth': 10, 'n_estimators': 150}   
0              10                100  {'max_depth': 10, 'n_estimators': 100}   
2              10                250  {'max_depth': 10, 'n_estimators': 250}   

   rank_test_score  split0_test_score  split0_train_score       ...         \
5                1           0.923248            1.000000       ...          
8                2           0.926585            1.000000       ...          
7                3           0.920282            1.000000       ...          
4                4           0.921765            1.000000       ...          
3                5           0.921023            1.000000       ...          
6                6           0.914720            1.000000       ...          
1                7           0.855024            0.913306       ...          
0                8           0.854653            0.909875       ...          
2                9           0.851687            0.911729       ...          

   split2_test_score  split2_train_score  split3_test_score  \
5           0.936225            1.000000           0.924703   
8           0.930293            1.000000           0.926558   
7           0.932888            1.000000           0.926187   
4           0.932518            1.000000           0.928412   
3           0.925844            1.000000           0.926929   
6           0.928810            1.000000           0.923591   
1           0.858361            0.915253           0.859050   
0           0.859103            0.914418           0.856083   
2           0.859844            0.914975           0.861276   

   split3_train_score  split4_test_score  split4_train_score  std_fit_time  \
5            1.000000           0.924304            1.000000      0.025922   
8            1.000000           0.923191            1.000000      0.507069   
7            1.000000           0.925046            1.000000      0.021690   
4            1.000000           0.921336            1.000000      0.011175   
3            1.000000           0.920594            1.000000      0.028452   
6            1.000000           0.922820            1.000000      0.012831   
1            0.913963           0.868646            0.917215      0.012252   
0            0.911737           0.866790            0.911653      0.320630   
2            0.916002           0.860853            0.912858      0.055095   

   std_score_time  std_test_score  std_train_score  
5        0.003464        0.006058         0.000000  
8        0.021802        0.003938         0.000000  
7        0.000335        0.004421         0.000000  
4        0.003975        0.004590         0.000000  
3        0.003061      

In [31]:
testnums = [0, 1, 2]

# select param_n_estimators as 100
# vary maximum depth by 10, 30, 50.
params = [
    [100, 50],
    [100, 30],
    [100, 10]
]

for test in testnums:
  random_forest = RandomForestClassifier(n_estimators=params[test][0], max_depth=params[test][1],
                             random_state=7, max_features = .33 ,oob_score=True)
  random_forest.fit(X_train, np.ravel(y_train))

  print("Accuracy:")
  print(random_forest.score(X_test, y_test))

  print("Importances:")
  print(sorted(zip(random_forest.feature_importances_, X_test.columns), reverse=True))
  
  print("Compliment OOB Error")
  print(random_forest.oob_score_)
  
  print("Confusion Matrix")
   
  cmat = confusion_matrix(y_test, random_forest.predict(X_test))
  mats = pd.DataFrame(cmat)
  print(mats)
  
  print("Sensitivity")
  print(mats.iloc[1,1]/( mats.iloc[1,1]+ mats.iloc[0,1]))
  print("Specificity")
  print(mats.iloc[0,0]/( mats.iloc[0,0]+ mats.iloc[0,1]))

Accuracy:
0.9365821094793058
Importances:
[(0.11826913374097234, 'O1'), (0.1114983410596552, 'P7'), (0.09662845393132381, 'F7'), (0.08105068401709646, 'F8'), (0.07771734236517845, 'AF4'), (0.07740391558668333, 'AF3'), (0.06137318233383824, 'FC6'), (0.06031666494886239, 'F4'), (0.05750537608606013, 'FC5'), (0.056916402905140634, 'T8'), (0.05649715424379695, 'O2'), (0.051991356357023726, 'T7'), (0.051900442605435304, 'F3'), (0.040931549818933, 'P8'), (0.0, 'Intercept')]
Compliment OOB Error
0.9283489096573209
Confusion Matrix
     0    1
0  768   40
1   55  635
Sensitivity
0.9407407407407408
Specificity
0.9504950495049505
Accuracy:
0.9345794392523364
Importances:
[(0.11834767817586986, 'O1'), (0.11142917776777957, 'P7'), (0.09717004242781796, 'F7'), (0.0810680937112625, 'F8'), (0.07771158215021752, 'AF4'), (0.07726353013931621, 'AF3'), (0.06119683920376835, 'FC6'), (0.06050418344310365, 'F4'), (0.057265909305491185, 'FC5'), (0.05696667721236987, 'T8'), (0.056540770398339296, 'O2'), (0.05

In [34]:
ran_forest = RandomForestClassifier(n_estimators=100, max_depth=50,
                             random_state=7, max_features = .33 ,oob_score=True)
ran_forest.fit(X_train, np.ravel(y_train))


print("Accuracy:")
print(ran_forest.score(X_test, y_test))

print("Importances:")
print(sorted(zip(ran_forest.feature_importances_, X_test.columns), reverse=True))
  
print("Compliment OOB Error")
print(ran_forest.oob_score_)
  
print("Confusion Matrix")
   
cmat = confusion_matrix(y_test, ran_forest.predict(X_test))
mats = pd.DataFrame(cmat)
print(mats)
  
print("Sensitivity")
print(mats.iloc[1,1]/( mats.iloc[1,1]+ mats.iloc[0,1]))
print("Specificity")
print(mats.iloc[0,0]/( mats.iloc[0,0]+ mats.iloc[0,1]))

Accuracy:
0.9365821094793058
Importances:
[(0.11826913374097234, 'O1'), (0.1114983410596552, 'P7'), (0.09662845393132381, 'F7'), (0.08105068401709646, 'F8'), (0.07771734236517845, 'AF4'), (0.07740391558668333, 'AF3'), (0.06137318233383824, 'FC6'), (0.06031666494886239, 'F4'), (0.05750537608606013, 'FC5'), (0.056916402905140634, 'T8'), (0.05649715424379695, 'O2'), (0.051991356357023726, 'T7'), (0.051900442605435304, 'F3'), (0.040931549818933, 'P8'), (0.0, 'Intercept')]
Compliment OOB Error
0.9283489096573209
Confusion Matrix
     0    1
0  768   40
1   55  635
Sensitivity
0.9407407407407408
Specificity
0.9504950495049505


## Summary

We notice that the out of bag score for the random forest model is acutely high > 86%. This result suggests 14% and greater of the predicted values in the "out of sample" data (unused in in the creation of the tree), is accurate. Furthermore, since our values of .oob_score_ fall way beyond 50% (by 46) our ensemble outperforms just flipping a coin by more than 46%, 

Furthermore, using an alternative Accuracy scoring method, our best gridsearch cross validation metrics indicate high accuracy with a maximum depth of 50 (92%). The result indicates a high percentage of correct prediction out of total number of predicted values. This fold had the lowest estimate of overfitting compared to all other cross validated random forest models.

Our five most important indicators include: 'O1', 'P7', 'F7', 'F8', and 'AF4'.